### PR TITLE
Brandon edwards default prob bug fix

### PR DESCRIPTION
**Pull request by brandon-edwards**
Bug fix: Set the default values for probabilities of words not seen in training (conditioned on a given topic) and topics (conditioned on documents not seen in training) to be zero.

Since final scoring is performed in one location for dos and proxy (SuspiciousConnectsScoreFunction.scala) and another for net flow (FlowPostLDA.scala); this fix required changes to two separate files. The previously set probabilities were 1/#topics and 0.05 respectively for these files. These values were changed to 0.

### CODE REVIEW

**Review by NathanSegerlind — APPROVED** [review]

File: spot-ml/src/main/scala/org/apache/spot/SuspiciousConnectsScoreFunction.scala [hunk 1]
```diff
@@ -11,7 +11,7 @@ class SuspiciousConnectsScoreFunction(topicCount: Int,
   def score(ip: String, word: String): Double = {
 
     val uniformProb = Array.fill(topicCount) {
-      1.0d / topicCount
+      0.0d / topicCount
```
Comment (NathanSegerlind) [inline spot-ml/src/main/scala/org/apache/spot/SuspiciousConnectsScoreFunction.scala]: why divide by topicCount and not just use 0d?

File: spot-ml/src/main/scala/org/apache/spot/SuspiciousConnectsScoreFunction.scala [hunk 2]
```diff
@@ -11,7 +11,7 @@ class SuspiciousConnectsScoreFunction(topicCount: Int,
   def score(ip: String, word: String): Double = {
 
     val uniformProb = Array.fill(topicCount) {
```
Comment (NathanSegerlind) [inline spot-ml/src/main/scala/org/apache/spot/SuspiciousConnectsScoreFunction.scala]: maybe "zeroProb" is a more appropriate name than uniformProb

File: spot-ml/src/main/scala/org/apache/spot/netflow/FlowPostLDA.scala [hunk 3]
```diff
@@ -99,7 +99,7 @@ object FlowPostLDA {
     val dataWithIpProb = dataWithIpProbJoin.selectExpr(newSchemaColumns: _*)
 
     def scoreFunction(word: String, ipProbabilities: Seq[Double], topicCount: Int): Double = {
-      val uniformProb = Array.fill(topicCount)(0.05d)
+      val uniformProb = Array.fill(topicCount)(0.0d)
```
Comment (NathanSegerlind) [inline spot-ml/src/main/scala/org/apache/spot/netflow/FlowPostLDA.scala]: zeroProb seems like a more appropriate name

**Comment by brandon-edwards** [comment]
I've committed new changes to fix the order of parameters being passed to FlowWordCreator in FlowSuspiciousConnectsModel.scala . I have now tested this code with all changes and get good results.

File: spot-ml/src/main/scala/org/apache/spot/netflow/model/FlowSuspiciousConnectsModel.scala [hunk 4]
```diff
@@ -58,13 +58,17 @@ class FlowSuspiciousConnectsModel(topicCount: Int,
     val wordToPerTopicProbBC = sc.broadcast(wordToPerTopicProb)
 
 
+    /** A left outer join (below) takes rows from the left DF for which the join expression is not
+      * satisfied (for any entry in the right DF), and fills in 'null' values (for the additional columns).
+      */
     val dataWithSrcTopicMix = {
-      val joinedDF = inDF.join(ipToTopicMixDF, inDF(SourceIP) === ipToTopicMixDF("ip"))
+      val joinedDF = inDF.join(ipToTopicMixDF, inDF(SourceIP) === ipToTopicMixDF("ip"), "left_outer")
```
Comment (NathanSegerlind) [inline spot-ml/src/main/scala/org/apache/spot/netflow/model/FlowSuspiciousConnectsModel.scala]: I think I named these variables, so it's not your fault Brandon...
but these variable names stink.   I am sorry. A variable name should tell us  what something is, not how it was computed (as that is redundant with the code... that is, "joinedDF is the dataframe you get by taking in the in dataframe and joining it with something else"

I suggest:

recordsWithSrcIPTopicMixes for joinedDF

and then

recordsWithTopicMixes for joinedDF2

sorry about that

File: spot-ml/src/main/scala/org/apache/spot/netflow/model/FlowScoreFunction.scala [hunk 5]
```diff
@@ -55,23 +55,29 @@ class FlowScoreFunction(timeCuts: Array[Double],
...
 
+    val FlowWords(srcWord, dstWord) = flowWordCreator.flowWords(hour: Int, minute: Int, second: Int,
+      srcPort: Int, dstPort: Int, ipkt: Long, ibyt: Long)
 
+    val zeroProb = Array.fill(topicCount) { 0.0 }
 
-    val uniformProb = Array.fill(topicCount) {
-      1.0d / topicCount
-    }
+    /** A null value for srcTopicMix or dstTopicMix indicated the ip (source or dest respectively)
```
Comment (rabarona) [inline spot-ml/src/main/scala/org/apache/spot/netflow/model/FlowScoreFunction.scala:64]: This comment is very helpful, maybe it should appear in SuspiciousConnectsScoreFunction.scala as well.

File: spot-ml/src/main/scala/org/apache/spot/netflow/model/FlowSuspiciousConnectsModel.scala (hunk 4 above)
Comment (rabarona) [reply spot-ml/src/main/scala/org/apache/spot/netflow/model/FlowSuspiciousConnectsModel.scala]: I agree with the variable names changes.

File: spot-ml/src/main/scala/org/apache/spot/netflow/model/FlowSuspiciousConnectsModel.scala [hunk 6]
```diff
@@ -184,7 +188,7 @@ object FlowSuspiciousConnectsModel {
 
     // simplify DNS log entries into "words"
 
-    val flowWordCreator = new FlowWordCreator(ibytCuts, ipktCuts, timeCuts)
+    val flowWordCreator = new FlowWordCreator(timeCuts, ibytCuts, ipktCuts)
```
Comment (rabarona) [inline spot-ml/src/main/scala/org/apache/spot/netflow/model/FlowSuspiciousConnectsModel.scala:192]: Was this an existing error? Have we been passing the wrong parameters to FlowWordCreator all the time? Or did FlowWordCreator constructor changed?

**Review by rabarona — CHANGES_REQUESTED** [review]

File: spot-ml/src/main/scala/org/apache/spot/netflow/model/FlowSuspiciousConnectsModel.scala (hunk 6 above)
Comment (NathanSegerlind) [reply spot-ml/src/main/scala/org/apache/spot/netflow/model/FlowSuspiciousConnectsModel.scala:192]: it was an existing error, but only in like the last few days on dev

Comment (NathanSegerlind) [reply spot-ml/src/main/scala/org/apache/spot/netflow/model/FlowSuspiciousConnectsModel.scala:192]: and only for netflow

File: spot-ml/src/main/scala/org/apache/spot/netflow/model/FlowScoreFunction.scala (hunk 5 above)
Comment (brandon-edwards) [reply spot-ml/src/main/scala/org/apache/spot/netflow/model/FlowScoreFunction.scala:64]: Yes, I agree, I'll put that in.

Comment (brandon-edwards) [reply spot-ml/src/main/scala/org/apache/spot/netflow/model/FlowScoreFunction.scala:64]: Done

**Review by rabarona — APPROVED** [review]

**Pull request merged**